### ISSUE TITLE
Closures in PSR2 are not checked for a space after the function keyword

### ISSUE THREAD

**Issue by ttyieu** [squizlabs/PHP_CodeSniffer#496]
Currently phpcs doesn't check for a space between the function keyword and argument list in closures, so following code doesn't raise any warning:
```php
$callback = function() {
};
```
I'm trying to follow PSR-2 and it's somewhat annoying because i make that mistake rather often.

Also, this code passes without any warning:
```php
$callback = function () use ($this){
};
```
Although it is not explicitly said that there must be a space after closing parenthesis of use list, it remains inconsistent with following case, for which there isn't an explicit rule either yet it actually raises an error:
```php
$callback = function (){
};
```

**Comment by aik099** [comment]
Can you please post a link to exact place in PSR2 standard, that says about that spaces.

I'm asking, because there were a few bug reports in past, that turned out to be misinterpretations of what standard actually defines.

**Comment by ttyieu** [comment]
https://github.com/php-fig/fig-standards/blob/821ea5c5e1a30823f83e1bd56118861ded9ebbc6/accepted/PSR-2-coding-style-guide.md#6-closures
First paragraph.
And as I said, it doesn't explicitly states that there must be a space before opening brace (only implies it later in the examples), but I think it's a shortcoming.

**Comment by gsherwood** [comment]
In this case, because PSR2 says to `Note the placement of parentheses, commas, spaces, and braces`, these spaces should be considered part of the standard and PHPCS should be checking them. I will mark this as a bug and add the checks.

Having said that, PSR2 wording is argued over all the time, so I fully expect someone to get new errors after this change and dispute it, or for people to be using newlines instead of spaces and so to have unexpected results.

Still, I'm putting this comment here so my position is clear. Thanks for reporting the issue.

**Comment by gsherwood** [comment]
This should be fixed now.

**Comment by mhujer** [comment]
@gsherwood It seems that the ee028f330cac33c54ac5df6b33b62460831aba39 breaks for closures when used as a parameter:

Generic.Functions.OpeningFunctionBraceBsdAllman.BraceOnSameLine *"Opening brace should be on a new line"*

```
        array_walk($data, function (&$item) {
            //few lines
        });
```

**Comment by gsherwood** [comment]
The two sniffs that check brace position were not supporting closures, when they should be.

Do you have different rules for closures and regular functions, and is this why you think the sniff is now broken?

**Comment by mhujer** [comment]
Thanks for the answer, let me explain further:

- our app is Zend Framework based, which runs on PHP 5.2, so its standard does not define formatting of closures (PHP 5.3+)
- our app is running on PHP 5.5, so we are using closures 
- our ruleset is based on Zend standard from PHPCS (which contains also the Generic.Functions.OpeningFunctionBraceBsdAllman rule)
- since the ee028f330cac33c54ac5df6b33b62460831aba39, closures in our codebase (formatted as in the previous comment) are reported as errors: *Opening brace should be on a new line*

this satisfies the PHPCS:
```
array_walk($data, function (&$item)
{      <--- opening brace on a new line
    //few lines of code here
});
```

So the question is - what is the correct placement of the opening brace of a closure. Opening brace of a methods seems natural on a new line, but an opening brace of a closure kinda fits in the same line as the function declaration.

**Comment by gsherwood** [comment]
> So the question is - what is the correct placement of the opening brace of a closure.

The Squiz, PEAR and PSR2 standards all keep the brace on the same line for closures. To implement this, and all other function/closure formatting checks, they use either the PEAR.Functions.FunctionDeclaration sniff or the Squiz.Functions.MultiLineFunctionDeclaration sniff (for a couple more formatting checks).

What I will do thought is try and make closure support optional in those sniffs, so they can be turned off in a ruleset file.

Also note that the Zend standard in PHPCS is not actually a real standard. It's a collection of sniffs contributed by someone at Zend many many years ago. Eventually (not before PHPCS v3) I'll remove it and move the sniffs into the Generic standard.

**Comment by gsherwood** [comment]
I've just committed a change that reverts the functionality back to how it was (checking functions only) but allow the sniffs to optionally support checking closures as well. Commit is here: https://github.com/squizlabs/PHP_CodeSniffer/commit/2688fe60200d6aa6d5c60b8c7a818c5cac60a772

**Comment by mhujer** [comment]
:+1: I can confirm that it works as previously with our original standard.

(thanks for the advise to move away from the Zend standard - I've replaced it with separate sniffs it contained. And the Squiz.Functions.MultiLineFunctionDeclaration found few more mistakes in our codebase)

**Comment by gsherwood** [comment]
> I can confirm that it works as previously with our original standard.

Thanks a lot for testing this for me.
